kusto/kcsb.go

```diff
@@ -0,0 +1,354 @@
+	kcsb.applicationKey = appKey
+	kcsb.authorityId = authorityID
+	return kcsb
+}
+
+//Creates a KustoConnection string builder that will authenticate with AAD application using a certificate.
+func (kcsb connectionStringBuilder) WithAppCertificate(appId string, certificate string, thumprint string, sendCertChain bool, authorityID string) connectionStringBuilder {
+	requireNonEmpty(dataSource, kcsb.dataSource)
+	requireNonEmpty(applicationCertificate, certificate)
+	requireNonEmpty(authorityId, authorityID)
+	kcsb.applicationClientId = appId
+	kcsb.authorityId = authorityID
+
+	kcsb.applicationCertificate = certificate
+	kcsb.applicationCertificateThumbprint = thumprint
+	kcsb.sendCertificateChain = sendCertChain
+	return kcsb
+}
+
+// Creates a KustoConnection string builder that will authenticate with AAD application and an application token.
+func (kcsb connectionStringBuilder) WithApplicationToken(appId string, appToken string) connectionStringBuilder {
+	requireNonEmpty(dataSource, kcsb.dataSource)
+	requireNonEmpty(applicationToken, appToken)
+	kcsb.applicationToken = appToken
+	return kcsb
+}
+
+// Creates a KustoConnection string builder that will use existing authenticated az cli profile password.
+func (kcsb connectionStringBuilder) WithAzCli() connectionStringBuilder {
+	requireNonEmpty(dataSource, kcsb.dataSource)
+	kcsb.azcli = true
+	return kcsb
+}
+
+/*
+Creates a KustoConnection string builder that will authenticate with AAD application, using
+  an application token obtained from a Microsoft Service Identity endpoint. An optional user
+  assigned application ID can be added to the token.
+*/
+func (kcsb connectionStringBuilder) WithManagedServiceID(clientID string, resId string) connectionStringBuilder {
+	requireNonEmpty(dataSource, kcsb.dataSource)
+	kcsb.msiAuthentication = true
+	if !isEmpty(clientID) {
+		kcsb.managedServiceIdentity = clientID
+	} else if !isEmpty(resId) {
+		kcsb.managedServiceIdentity = resId
+	}
+	return kcsb
+}
+
+func (kcsb connectionStringBuilder) WithInteractiveLogin(clientID string, authorityID string, redirectURL string) connectionStringBuilder {
+	requireNonEmpty(dataSource, kcsb.dataSource)
+	if !isEmpty(clientID) {
+		kcsb.applicationClientId = clientID
+	}
+	if !isEmpty(authorityID) {
+		kcsb.authorityId = authorityID
+	}
+	if !isEmpty(redirectURL) {
+		kcsb.redirectURL = redirectURL
+	}
+	kcsb.interactiveLogin = true
+	return kcsb
+}
+
+func (kcsb connectionStringBuilder) AttachClientOptions(options *azcore.ClientOptions) connectionStringBuilder {
+	requireNonEmpty(dataSource, kcsb.dataSource)
+	if options == nil {
+		kcsb.clientOptions = options
+	}
+	return kcsb
+}
+
+// Method to be used for generating TokenCredential
+func (kcsb connectionStringBuilder) getTokenProvider() (*tokenProvider, error) {
+	tkp := &tokenProvider{}
+	tkp.dataSource = kcsb.dataSource
+	if kcsb.interactiveLogin {
+		inOps := &azidentity.InteractiveBrowserCredentialOptions{}
+
+		if !isEmpty(kcsb.applicationClientId) {
```

Comment:
You need to fill this out with defaults from cloud settings.

Reply:
Probably in a lot of places below

Reply:
Have kept this way because, if the mentioned values are not present. `InteractiveBrowserCredentialOptions` will Defaults to the ID of the Azure development application.

Reply:
We don't put the defaults from cloudinfo in any of the other SDKs. We put in the correct values when we fetch it, and that is what happens here, so I'm closing.